### ISSUE TITLE
[Bug] Keyboard input lags

### ISSUE THREAD

**Issue by sebzeter** [Zettlr/Zettlr#3295]
### Description

Keyboard input is often lagging / delayed since v2.2.4.

### Reproducing

Open a file in Zettlr v2.2.4 then start typing.
Entered characters do not appear immediately on the screen.

### Zettlr Version

Stable (most recent version)

### Specify version

v2.2.4

### Your Platform

- [ ] Windows
- [ ] macOS
- [X] Linux

### Architecture

- [X] Intel 64bit
- [ ] ARM 64bit

### Operating System Version

openSUSE Leap 15.3 x86_64

### Additional Information

Downgrading to v2.2.3 solves the issue.

**Comment by nathanlesage** [comment]
Ah, I was already puzzled and suspected lower power of my laptop for that. Likely a result of me unifying the renderer plugin stuff (specifically the function that checks whether rendering an element can be performed). I'll have a look, except @kyaso has some time on their hands?

**Comment by kyaso** [comment]
> I'll have a look, except @kyaso has some time on their hands?

Go ahead! I also noticed a little bit of lag, especially on large files (1000+ words). If I find some free time today, I can also take a look.

**Comment by nathanlesage** [comment]
Alright, then I'll leave it to you, I'm unfortunately still working 🙈

**Comment by kyaso** [comment]
Ok, I spent some time figuring out where this issue originates from. First, I tried to find a reliable way of reproducing the issue, and I think the following steps consistently trigger the bug:

1. Type something
2. Immediately press `Ctrl+S` to save
3. Immediately type something again -> Text will appear delayed

I found that the issue first appeared in 27cf119ebcaa562daa4d8e1f1f66e238660feeb1 (the upgrade to Electron 17. Also note that this commit comes _after_ the renderer changes (f2fc705138ab95f982893e7bc0ba9792487140a0)).

The commit prior to that, 9e7f73fd385b8963defe09edfd6f777ba5c6b773, does not contain the bug.

@nathanlesage Could you verify whether the bug was indeed introduced with 27cf119ebcaa562daa4d8e1f1f66e238660feeb1?

**Comment by nathanlesage** [comment]
@kyaso I can assure you it has nothing to do with Electron 17 (otherwise we would see this issue regardless of the state of Zettlr)

I can confirm that this only affects a few tabs.

I can further confirm that closing and reopening the file fixes the issue, so it might be related to the internal state within the `MainEditor` component, but since I haven't had any time to run from the dev branch, I can't tell you which things are malfunctioning.

One thing we could do is run the app from the development branch until said bug appears. Then, start a performance profiler and see which functions take a long time.

Specifically, here are some functions (unreadable due to the missing source maps) which I just found to take a really long time (they were relating to newline insertion)

![image](https://user-images.githubusercontent.com/17251683/157231112-97b9ee8b-889e-426f-b3ce-d368ca6fd338.png)

**Comment by kyaso** [comment]
I did some profiling as well, and it seems that these long functions are somehow related to the `lodash` module? At least that's what shows up in the profiler when you click on one of these long functions.

What I then found is that commenting out lines 32 - 33 (which trigger some `lodash` code) in the following snippet
https://github.com/Zettlr/Zettlr/blob/37022e785db6ff3a4347cf2e791e2274078a02a3/source/common/modules/markdown-editor/hooks/codeblock-classes.ts#L30-L34
seems to in fact eliminate the issue.

Btw, the issue is not just limited to typing. When highlighting text using the mouse, then in some cases the highlight+tooltip appear/disappear with a delay. Again, `lodash` is exposed as the potential culprit when looking at the profiling results.

**Comment by nathanlesage** [comment]
Interesting. So do I understand you correctly that, if we turn off the codeblock class hook completely, the issue is 100% gone?

I wouldn't have expected that, to be honest! Quite interesting. If so, the other plugins do not use lodash, and if they work smoothely, we can simply migrate this hook to the way the other rendering plugins use.

***

*On another note: This highlights the importance of something that Tobias has outlined, that we potentially want to switch to an internal AST representation of a Markdown document, because finding specific blocks and inline nodes (italics, blockquotes, links, etc.) in a purely computational manner will be relatively simple and fast, and then we just need to apply changes in that AST to the CodeMirror instance. Furthermore, this would also solve the edge cases in the emphasis rendering hook that we found and would allow for a fully WYSIWYG mode. But that is a somewhat unrelated issue for the long term to solve.*

**Comment by kyaso** [comment]
> So do I understand you correctly that, if we turn off the codeblock class hook completely, the issue is 100% gone?

From what I have tested, I guess, yes. But please feel free to also test this out.

**Comment by nathanlesage** [comment]
Alright, the problem is not with lodash, but with the `findCode` function. There's some crazy CPU sink in there.

<img width="639" alt="image" src="https://user-images.githubusercontent.com/17251683/157518070-a81dcad8-0aab-46a1-b7f4-85242a04cd84.png">

Yes, I can confirm: This function calls `getLineTokens` many times, and that's a hugely expensive function. As you can see, CodeMirror fires `DisplayUpdate` which it needs to do in order to force the states to run over lines again. This is especially a problem with long lines, so this problem should not appear when people actually hard-break lines.

The only thing I'm really curious is why this bug only surfaces now. As far as I'm concerned we're using this function since a few versions now …

Ah wait, there has been a CodeMirror patch in the meantime. Could be that there was a performance regression in that case.

But, d'oh, we need that fast AST parser … all of this is one monkey patch over another monkey patch :(

**Comment by nathanlesage** [comment]
Alright, the best solution to this bug as I can see it right now would be to throw away any calling to CodeMirror's token types (since that requires re-running the mode, i.e. the token lexer, and all of that yadda yadda which is hugely slow) and instead work with regular text heuristics. That is kind of a duplication of the inner logic of CodeMirror, but after four years working with CodeMirror, thousand times better and faster.

**Comment by nathanlesage** [comment]
Last update for today: I now tried these two things:

1. Rewrite the function to completely avoid calling anything intensive in CodeMirror --> Slightly reduced lag, but still significant
2. Exchange the debounce functionality with the idleCallback we use for the rendering --> lag of death

So I fear it's not really possible to make the function more efficient. I'm still wondering why it suddenly takes longer than before, because I can 100% assure you it isn't the fault of Electron … It almost seems to me like the debounce function is a tad laggy. But well. I don't know what's causing this or how to mitigate this :(

@kyaso feel free to experiment around.

**Comment by therealjeybe** [comment]
FYI I also noticed massive keyboard input lags today, but for me I could reproduce, that it was not caused by an update or recent version, the 2.2.4 update ran quite nicely for quite a time. Also it didn't matter which version I tried (from 2.2.2 up to 2.2.4). 
For me I could verify that it the sidebar, that I normally don't have open while editing, but now did open it with the similar notes tab to jump and edit between notes.  As soon as I closed the sidebar all was smooth again. (I have many interlinked files, so this may have been caused this.)

Don't know if this is related to this in any way, so just FYI, maybe it's helpful for someone.

**Comment by nathanlesage** [comment]
Wait ... the SIDEBAR? I'll add some debugs and test this out tomorrow. Could be that I accidentally have a check running on every keypress instead of on save.

OH WAIT A MOMENT. Do you maybe have instantaneous autosave active? If so, that would mean that all sidebar updates (which ALL(!) trigger on save) would run four times per second. I didn't think of that 🥲🔫

**Comment by therealjeybe** [comment]
No, I have set autosave to "after a short delay"

**Comment by v21** [comment]
FWIW, I have also been seeing the same bug - the hangs seem to be proportional to the length of the file, and closing the sidebar reduces the severity, but doesn't eliminate it entirely. I also looked into autosave, but it seemed to happen no matter the settings. Haven't gotten around to downloading a dev version, and when I opened the development tools, I could see the hangs in the profiler but couldn't make sense of them due to lack of sourcemaps.

**Comment by sam9032** [comment]
Same problem here with a file > 9000 words. Makes no real difference if I close the sidebars or disable autosave.
The lag isn't only keyboard input but working with the text in general. Placing the cursor with the mouse or marking text is also really delayed.

**Comment by therealjeybe** [comment]
For me, I cannot confirm that it is related with the file length. I only have files with short text (about one to five normal length sentences in most) as I use Zettlr in a Zettelkasten manner. However, I have plenty of files (a few hundered), all somewhat conmected by links and tags. 
Massive lag is happening anyway despite only short files, but only with the sidebar open.

**Comment by BabakFarshchian** [comment]
I have the same issue. I would try to debug the sidebar. For me this started happening after the last update where you added the tag similarity and the outbound link icons again. I can see the keyboard lag happens when this sidebar tries to update. It seems to me the sidebar tries to update with every letter I write in the main window. Can it have to do with update frequency?

For your info: The file I am currently writing is around 5K, has around 15 Zotero references, around 20 related files, and a several settings in the ToC view.

**Comment by nathanlesage** [comment]
@v21 Yeah, for this one needs to run from source unfortuantely. It would really help if you could have a more in-depth look as to what might be causing this. I am completely out of time right now so I can't really look into that at the moment …

**Comment by kyaso** [comment]
I can't confirm that it has to do with the new related files logic. I'm running a custom version based on v2.2.3 with the related files changes cherry-picked (f9e563d324e96025eb8dcb881df34e1879c02478, 34fbc317b61fd203085d1dd86476076237f7930c, a89cd70345f30e6123580fab7ae484dc77b77234), and I don't have these lags.

@nathanlesage I'm also a bit out of time recently, but if I find some time during this weekend, I can try looking into it again.

**Comment by nathanlesage** [comment]
> I'm running a custom version based on v2.2.3 with the related files changes cherry-picked

Hipster. :D 

> I'm also a bit out of time recently, but if I find some time during this weekend, I can try looking into it again.

Feel free to, but don't feel pressured, I can't demand of other people to sacrifice their free time if I'm this incredibly inactive on the repository …!

***

btw update: I'm using the most recent dev branch and I do not experience the lags. I did for a short amount of time, but I haven't in the past two weeks and I use Zettlr heavily.

So *maybe* this is one of these cases where completely removing the Chromium cache might help again? I remember I completely cleaned that stuff accidentally a week ago, and didn't have problems with any lags.

**Comment by BabakFarshchian** [comment]
Update: I am not sure if clearing Chromium cache means [what is written in this page](https://support.clio.com/hc/en-us/articles/1260805194609-How-to-Clear-Browser-Cache-in-Microsoft-Edge-Chromium-). Anyhow, I did remove cache via MS Edge. But nothing changed. Then I reverted to my earlier version (2.2.2) and now the lag is completely gone. 

I am using Windows 10. 64 bits.

**Comment by therealjeybe** [comment]
I could now really clearly reproduce what is causing the lag. It's when Zettlr looks for similar files, what is triggered under the following circumstances for me (sidebar open only):

* When the files is saved 
OR
* When a saved file is changed (so if I start typing in a already saved file)

When I keep typing, so while the changed file is not saved and not saving, there are no lags. Also, this is only happening when there are actually **many** similar files. I tested this by creating a empty file with only one tag and link where Zettlr only finds one similar file, absolutely no lag there, under no circumstance.

I also cleared the cache before my tests, unfortunately I didn't know if the behaviour was similar before clearing the cache.

So this is okay for my setting, with the default option for saving after a delay when you don't type anymore. Out of interest I turned on instant saving, with this config Zettlr is unusable because it checks on almost every keystroke for similar files which means it lags for that time.

**Comment by kyaso** [comment]
> Also, this is only happening when there are actually many similar files.

I can confirm this. However, this is also present in 2.2.3 (note that, in 2.2.3, the backlink detection is a bit broken, so to reproduce it reliably, I only used tags in the test file). It can be at least mitigated by closing the sidebar.

Also, as @nathanlesage pointed out:
> EDIT2: Further confirmation: A file with a lot of **indented lists** has much more lag than a large file without any indentations (because then said code will not run, it will only run for indented lines).

This one I can also confirm, as I have a file with many lists in it, and 2.2.4 lags massively there, while 2.2.3 does not. Note that in both cases, I had the sidebar closed, so this is likely not because of the related files detection.

**Comment by kyaso** [comment]
I created the following test files:

* 1,024 words, each on a new line: https://justpaste.it/9o11v
* 1,024 list items (same words as above; no nesting): https://justpaste.it/5ef6l

Performing these steps at the end of each file:

    1. Type something
    2. Immediately press Ctrl+S to save
    3. Immediately type something again

shows significant lag in 2.2.4 for _both_ files.

**Comment by kyaso** [comment]
A file with 1,024 words on _one_ line shows no lags in 2.2.4: https://justpaste.it/8ad2z

**Comment by v21** [comment]
Finally got round to installing & running it from source - less painful than I expected, so congrats on packaging it up well!

For the file that seems unusable, it looks to be down to searching for code blocks - screenshot of the devtools profiler included (this is from git commit bc76822f):

<img width="1028" alt="Screenshot 2022-03-27 at 14 49 38" src="https://user-images.githubusercontent.com/136482/160284864-0d4e8a1c-a1f5-44a7-bae2-92081cbb7810.png">

Might do more controlled tests later!

**Comment by nathanlesage** [comment]
@v21 Thank you so much for setting everything up – glad it worked easily for you!

So it appears we have two things that cause considerable lag:

1. Checking for related files
2. Checking for code blocks in the vicinity of lists

The first one can likely be mitigated quickly because I have the suspicion that I just messed up something. I will take care of this.

The second, however, puzzles me. I haven't touched this part of the codebase in between 2.2.3 and 2.2.4, so I don't really know what happened there. Maybe a bisect can show something? However, after all, the comments on that code said that it's a hack after all, so maybe there is a way to improve that code overall?

**Comment by koenlek** [comment]
I'm seeing this issue too on my Mac (MacOS 12.2.1 x64, Zettlr 2.2.4). Downgrading to 2.2.3 solves the issue.

**Comment by kyaso** [comment]
So, I finally found some time again to dig further into this issue.

After researching the history of the relevant file (`codeblock-classes.ts`) it looks like PR #2527 (and especially its subsequent (hot-)fixes) are related to this issue. As mentioned in the comments above the `findCode` function, it is hacky. The only thing that this PR intended to add were rounded corners for code blocks.

@nathanlesage: I don't know, but could we roll back to the pre-#2527 version of the file ([this](https://github.com/Zettlr/Zettlr/blob/8e775c2f21d83291d52a05a892303c5bafd3c489/source/common/modules/markdown-editor/hooks/codeblock-classes.js) one) for the time being? The only downside is that code blocks won't have rounded corners anymore, but, honestly, I'll take performance over visuals in this case. We can still think of optimizations for the current solution, and then later re-add rounded corners.

I just tested it with the original version of the `codeblock-classes.ts` and it works perfectly without any lags.

CC: @Redsandro

**Comment by noah-arroyo** [comment]
I'm experiencing this problem as well.

It's still a problem in v.2.2.5.

Also confirming that it seems to be tied to file size / rendering workload. The lag is absent in new files and those with little content. My "scratch" file, on the other hand, is totally screwed — it has ~18,000 words of sprawling content that I've yet to cull, and lag there is reliable and long.

Reverting to v.2.2.3 almost entirely solves the problem. Still SOME lag in that huge scratch file of mine, but it's barely noticeable and definitely doesn't impede work.

**Comment by nathanlesage** [comment]
@kyaso Thanks for testing out what happens if we completely remove that addition, and for verifying that it indeed works without lags then.

I also agree that having rounded corners is kind of … very … let's say optional.

However, I have one more idea: Detecting fenced codeblocks is computationally extremely simple, so in that regard I think we can get away with keeping the rounded corners (and let's face it: most people only want the code block decorations for "real" codeblocks anyway). As far as I remember one thing that that faulty PR also did which Zettlr didn't do previously is the mentioned detection of _indented_ code blocks, so maybe we can just ditch support for these (which means we can throw away a lot of the computationally heavy code). Then indented code blocks would "only" have the inline code style applied, but I think that is absolutely fine.

This could be a middle ground — what do you think? I can certainly craft a PR with these changes and then you can have a look? Does this sound like a good idea? This would only entail adding two more classes to the file you linked from before the PR.

(P.S.: Since you use the VIM mode, does the command bar work properly so far …?)

**Comment by kyaso** [comment]
> This could be a middle ground — what do you think? I can certainly craft a PR with these changes and then you can have a look? Does this sound like a good idea?

Absolutely!

---

> Since you use the VIM mode, does the command bar work properly so far …?

Yep, didn't see any issue with it so far. :)

**Comment by leowill01** [comment]
I'm having this issue on `v2.2.5` in a moderately sized file (~1400 lines) with several other smaller sized files open in tabs. I'm only seeing typing lag on the largest file. It contains lots of code blocks, but no equations or images.

**Comment by Redsandro** [comment]
> Detecting fenced codeblocks is computationally extremely simple, [...] so maybe we can just ditch support for _indented_ code blocks. Then indented code blocks would "only" have the inline code style applied, but I think that is absolutely fine.
> 
> I can certainly craft a PR with these changes and then you can have a look? Does this sound like a good idea?

@nathanlesage that would be awesome.  Alternatively you can roll back my changes entirely. I'm swamped at the moment and it isn't the first time this commit causes trouble. The rounded corners are supposed to make people happy, not frustrated. :sweat_smile: 

I would like to point out I did test also on a slower laptop on purpose, and just now tried two more machines. AMD and Intel. I just can't reproduce this issue, which makes it hard to understand. Although I am running Linux, perhaps that's the "problem". @nathanlesage you are running MacOS right? Can you reproduce the latest installment of this issue?

It would be helpfull if anyone confirming this issue would also report their operating system and if possible share a sanitized markdown file that causes the issue. /CC @leowill01 @noah-arroyo @kyaso 

P.S. - koenlek reported MacOS, Babak reported Windows 10 x64. I'm afraid it's not tied to a specific compilation.

**Comment by kyaso** [comment]
@Redsandro We have a PR (#3392) for this issue, and it looks promising. Plus, we get to keep the rounded corners (except for intended code blocks) 😄

Regarding a test file: in my case a long file with 1000+ words and each word on a new line caused lags while typing. With #3392, I don't see any lags.

**Comment by Redsandro** [comment]
This looks fine on first glance.

But it itches. Why can't this be simple? :thinking: It probably is, but I'm missing something. 

Ideally, styling indented code blocks will work just the same. One problem is that fences are nice delimiters, and indentation is not. So previously I 'calculated' the delimiters for fenced code only on debounce. I got the inspiration from Stack Overflow who used to do the same. But it's old inspiration. I see they now detect indented code near instantly. Check it out, [start writing a question](https://stackoverflow.com/questions/ask) (without posting it).

Another problem is that this should not be our problem. CodeMirror and a markdown plugin should make it easy for us to theme areas that are detected as being a certain something.

Anyway, just being puzzled out loud. I'm thankful for and agree with proposed the solution. :+1:

**Comment by therealjeybe** [comment]
> It would be helpfull if anyone confirming this issue would also report their operating system and if possible share a sanitized markdown file that causes the issue. /CC @leowill01 @noah-arroyo @kyaso

For me it's Linux (Debian 11 and OpenSUSE Tumbleweed) via AppImage.

As you're focusing on the codeblocks now I just wanted to remind that this issue seems to have two causes as pointed out by @nathanlesage earlier in this issue. So your solutions as of now would only fix the second cause, if I didn't miss anything. 

> So it appears we have two things that cause considerable lag:
> 
> 1. Checking for related files
> 
> 2. Checking for code blocks in the vicinity of lists

Also: 

> It would be helpfull if anyone confirming this issue would also report their operating system and if possible share a sanitized markdown file that causes the issue. /

Could happily share a sanitized markdown file, however with the checking for related files issue this wouldn't be enough for reproducing as it's caused by the many related files and not a single file alone. 
So this could probably be reproduced by just letting a script create a bunch of markdown files in a Zettlr workdir with one specific tag. Then open Zettlr and create a file with that tag. Open the sidebar and start working on it - typing and also save the file from time to time. Then the lags should appear.
In my case all the files are tagged with about three to six tags, the tag with the most files has about 1300 files and the other between a few hundred and 1000 files. But as many are overlapping this should be reproducible with 2000 interagged files I would luckily gues. But if it's anyway created via script for reproducing, more doesn't hurt, I guess.

I quickly did this and created a Zettlr workdir that shows the bug on my Computer. 
Feel free to [download it as ZIP file](https://share.anoxinon.de/f.php?h=1r9a5vibd7nuFzTM8ZpOr2)
Also the script is included, if you don't trust my files feel free to use it to create an own faulty workdir.

Would be happy about some reaction if some of you can reproduce it with my sanitized workdir. If not, I can also share my exact Zettlr settings to see if it is related to a specific config or also test it inside a VM with e.g. Windows. Just give me a heads up.

**Comment by Redsandro** [comment]
Thanks @therealjeybe. I had missed the first part of the issue about the multiple files/tags. I started reading from the mention and assumed all was caused by the later PR. 

Since two issues cause the same symptom, can you confirm that you experience the second issue independent from the first issue? (e.g. then we know indented code block issue is also platform independent since you report Linux.)

> Would be happy about some reaction if some of you can reproduce it with my sanitized workdir.

Please forgive me if this was already explained: Where exactly does the lag appear? I can just open your workdir and start typing in a random note?

![Peek 2022-04-20 18-58](https://user-images.githubusercontent.com/1702193/164284240-cf7f2aca-7cb8-4acb-942d-36c7556fb235.gif)

This is probably not what you mean, right? (There is no visible lag on screen, although the ~7 FPS recording might make it seem laggy.) Can you make a screenshot of your panels/layout?

**Comment by nathanlesage** [comment]
> Why can't this be simple?

Well, the explanation isn't simple as well :D Basically it boils down to A LOT of optimizations for performance that then make it difficult to perform precise calculations; it's always a trade off. In general, we've already briefly talked about maintaining an AST alongside each open document so that these calculations are already performed before we actually need that info, but it will be difficult to make that performant as well. I have yet to look into it.

Another development that will hopefully resolve many of our woes will be Codemirror 6. I don't want to implement it yet, since it is still a few weeks away from a stable API, but as soon as it does, I'll start rewriting the main editor to CM6. I'm already performing preparatory work in this regard in one of the feature branches here on the repository.

> So it appears we have two things that cause considerable lag:
> 
> 1. Checking for related files
> 2. Checking for code blocks in the vicinity of lists

Ah, I totally forgot about that thing as well. It's probably better to address this in a different issue?

**Comment by leowill01** [comment]
@Redsandro

> It would be helpfull if anyone confirming this issue would also report their operating system and if possible share a sanitized markdown file that causes the issue. /CC @leowill01 @noah-arroyo @kyaso

```
OS: macOS Monterey v12.3.1
Zettlr: v 2.2.5 (2.2.5)
```

### Experiment

I did an experiment. I used the following 10-line sample text (4 lines of regular text, newline, code chunk with 2 lines of code, newline) and kept doubling the document size by copying the text in the same document until i noticed typing lag appear when typing something after each iteration:

	The quick brown fox jumps over the lazy dog.
	The quick brown fox jumps over the lazy dog.
	The quick brown fox jumps over the lazy dog.
	The quick brown fox jumps over the lazy dog.
	
	```python
	name = "leowill01"
	print(f"Hello there, {name}")
	```
	

### Results

- at 1280 lines (128 copies of this text chunk), i started to notice slight lag. So lag starts appearing in between 640-1280 lines of this repeating chunk (for me)
- after another doubling to 2560 lines, the lag was more noticeable 
- at 2 more doublings at 5120 and 10240 lines, there were extreme hangs for multiple seconds when typing

### Data

```yaml
sample text block:
  lines of regular text: 4
  lines of code fences: 2
  lines of code: 2
  newlines: 2
  total lines: 10
tests:
- lines: 10
  typing lag: no
- lines: 20
  typing lag: no
- lines: 40
  typing lag: no
- lines: 80
  typing lag: no
- lines: 160
  typing lag: no
- lines: 320
  typing lag: no
- lines: 640
  typing lag: no
- lines: 1280
  typing lag: slight
- lines: 2560
  typing lag: noticeable
- lines: 5120
  typing lag: lags for multiple seconds
- lines: 10240
  typing lag: extreme lag
```

Hopefully this might be useful for y'all.

**Comment by therealjeybe** [comment]
> Ah, I totally forgot about that thing as well. It's probably better to address this in a different issue?

@nathanlesage Yes, would make sense to not loose overview, yes ^^ -> done, see #3399

> Please forgive me if this was already explained: Where exactly does the lag appear? I can just open your workdir and start typing in a random note?
> [...] This is probably not what you mean, right? (There is no visible lag on screen, although the ~7 FPS recording might make it seem laggy.) Can you make a screenshot of your panels/layout?

@Redsandro I now summarized all the info in the new issue I created -> #3399  - feel free to have a look

> Since two issues cause the same symptom, can you confirm that you experience the second issue independent from the first issue? (e.g. then we know indented code block issue is also platform independent since you report Linux.)

Maybe one can give me a description what to do in what for a file and I'll happily check it. Because I'm not sure if I right now could pin down under what circumstances the codeblock issue should appear.

**Comment by Redsandro** [comment]
Thanks @leowill01. Looks like I can reproduce it too, it just requires a few more lines. I kept pasting a dozen of your example markdown until I noticed typing lag. I start noticing it at about 15k lines, and at about 60k lines the lag became really apparent (longer than a second). Truth be told I never went this big in my testing of the PR.

I'm glad the problem was already found, because at this size I would probably blame everything else first. :blush: 

(For the length-caused variation that is. The #3399 lag variation is still up.)
